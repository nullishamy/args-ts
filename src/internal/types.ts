import { Args } from '../args'

export type ExtractArgType<ArgObject, Default = never> = ArgObject extends Args<infer TArgs> ? TArgs : Default
