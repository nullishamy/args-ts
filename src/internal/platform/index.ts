export interface Platform {
  getEnv: (key: string) => string | undefined
  setEnv: (key: string, value: string) => void
  argv: () => string[]
  process: () => {
    versions: Record<string, string | undefined>
    isElectron: () => boolean
    isBundledElectron: () => boolean
    exit: (code: number) => never
  }
}

let _platform: Platform | undefined

export function usePlatform (platform: Platform): void {
  _platform = platform
}

export function currentPlatform (): Platform {
  if (!_platform) {
    throw new Error('no platform setup')
  }

  return _platform
}

export * from './node'
