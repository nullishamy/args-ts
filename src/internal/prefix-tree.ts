interface Node<T> {
  children: Record<string, Node<T> | undefined>
  value: T | undefined
  isTerminal: boolean
}

interface SearchHit<T> {
  didFind: true
  found: Node<T>
}

interface SearchMissed<T> {
  didFind: false
  previousNode: Node<T>
}

export class PrefixTree<V> {
  private readonly base: Record<string, Node<V> | undefined> = {}

  find (key: string): SearchHit<V> | SearchMissed<V> | undefined {
    let found = this.base[key[0]]

    if (!found) {
      return undefined
    }

    for (const char of key.substring(1)) {
      const next: Node<V> | undefined = found.children[char]

      if (next) {
        found = next
      } else {
        return {
          didFind: false,
          previousNode: found
        }
      }
    }

    return {
      didFind: true,
      found
    }
  }

  insert (key: string, value: V): void {
    let found = this.base[key[0]]
    if (!found) {
      found = {
        children: {},
        value: undefined,
        isTerminal: false
      }

      this.base[key[0]] = found
    }

    for (const char of key.substring(1)) {
      if (found.children[key] === undefined) {
        found.children[char] = {
          children: {},
          value: undefined,
          isTerminal: false
        }
      }

      found = found.children[char]
      if (!found) {
        throw new Error('impossible, no node found')
      }
    }

    found.value = value
    found.isTerminal = true
  }

  delete (key: string): Node<V> | undefined {
    const root = this.base[key[0]]
    if (!root) {
      return undefined
    }

    return this._delete(root, key)
  }

  toString (): string {
    return JSON.stringify(this.base, undefined, 2)
  }

  private _delete (node: Node<V>, key: string): Node<V> | undefined {
    if (key.length === 0) {
      if (node.isTerminal) {
        node.isTerminal = false
        node.value = undefined
      }

      if (node.children.size) {
        return node
      }

      return undefined
    }

    const newChild = node.children[key[0]]
    if (!newChild) {
      throw new Error(`no child for key ${key}`)
    }

    node.children[key[0]] = this._delete(newChild, key.substring(1))
  }
}
