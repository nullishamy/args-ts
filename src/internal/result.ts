interface ResultOk<T> {
  ok: true
  val: T
}

interface ResultErr<T> {
  ok: false
  err: T
}

export type Result<T, E = unknown> = ResultOk<T> | ResultErr<E>

export function Ok<T> (val: T): ResultOk<T> {
  const object = {
    ok: true,
    val
  } as const

  const stringify = (): string => `Ok(${JSON.stringify(val, undefined, 2)})`

  Object.defineProperty(object, 'toString', {
    value: stringify
  })

  return object
}

export function Err<E extends Error | Error[]> (err: E): ResultErr<E> {
  const object = {
    ok: false,
    err
  } as const

  const stringify = (error: Error): string => `Err(${error.constructor.name} { ${error.message} })`

  let chosenFunc
  if (Array.isArray(err)) {
    chosenFunc = () => {
      return err.map(stringify)
    }
  } else {
    chosenFunc = () => stringify(err)
  }

  Object.defineProperty(object, 'toString', {
    value: chosenFunc
  })

  return object
}
