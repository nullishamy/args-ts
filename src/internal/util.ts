import { InternalError, SchemaError } from '../error'
import { FlagAlias, InternalArgument } from './parse/types'

export function getArgDenotion (argument: InternalArgument): string {
  if (argument.type === 'flag') {
    return `--${argument.longFlag}`
  } else {
    return `<${argument.key}>`
  }
}

export function getAliasDenotion (alias: FlagAlias): string {
  if (alias.type === 'long') {
    return `--${alias.value}`
  } else {
    return `-${alias.value}`
  }
}

const flagValidationRegex = /-+(?:[a-zA-Z]+)/

export function internaliseFlagString (flag: string): ['long' | 'short', string] {
  if (!flagValidationRegex.test(flag)) {
    throw new SchemaError(`flags must match '--abcdefABCDEF' or '-abcdefABCDEF' got '${flag}'`)
  }

  // Long flag
  if (flag.startsWith('--')) {
    return ['long', flag.substring(2)]
  }

  if (flag.startsWith('-')) {
    return ['short', flag.substring(1)]
  }

  throw new InternalError('impossible')
}
