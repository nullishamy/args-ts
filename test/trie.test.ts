import assert from 'assert'
import { PrefixTree } from '../src/internal/prefix-tree'

describe('Trie tests', () => {
  it('can insert a small string', () => {
    const tree = new PrefixTree<string>()
    tree.insert('prefix', 'value')
    expect(tree).toMatchSnapshot()
  })

  it('can find a small string', () => {
    const tree = new PrefixTree<string>()
    tree.insert('prefix', 'value')

    const search = tree.find('prefix')
    assert(search !== undefined)
    assert(search.didFind)
    expect(search.found).toStrictEqual({
      children: {},
      value: 'value',
      isTerminal: true
    })
  })

  it('can find a string when it has a similar neighbour', () => {
    const tree = new PrefixTree<string>()
    tree.insert('prefix', 'value')
    tree.insert('prefax', 'value2')

    const search = tree.find('prefix')
    assert(search !== undefined)
    assert(search.didFind)
    expect(search.found).toStrictEqual({
      children: {},
      value: 'value',
      isTerminal: true
    })
  })

  it('can provide neighbours for failed searched', () => {
    const tree = new PrefixTree<string>()
    tree.insert('prefix', 'value')

    const search = tree.find('prefia')
    assert(search !== undefined)
    assert(!search.didFind)
    expect(search.previousNode).toStrictEqual({
      children: {
        x: {
          children: {},
          value: 'value',
          isTerminal: true
        }
      },
      value: undefined,
      isTerminal: false
    })
  })

  it('can provide the next values for a partial search', () => {
    const tree = new PrefixTree<string>()
    tree.insert('prefix', 'value')

    const search = tree.find('prefi')
    assert(search !== undefined)
    assert(search.didFind)
    expect(search.found).toStrictEqual({
      children: {
        x: {
          children: {},
          value: 'value',
          isTerminal: true
        }
      },
      value: undefined,
      isTerminal: false
    })
  })
})
