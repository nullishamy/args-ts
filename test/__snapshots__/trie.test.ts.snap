// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trie tests can insert a small string 1`] = `
PrefixTree {
  "base": {
    "p": {
      "children": {
        "r": {
          "children": {
            "e": {
              "children": {
                "f": {
                  "children": {
                    "i": {
                      "children": {
                        "x": {
                          "children": {},
                          "isTerminal": true,
                          "value": "value",
                        },
                      },
                      "isTerminal": false,
                      "value": undefined,
                    },
                  },
                  "isTerminal": false,
                  "value": undefined,
                },
              },
              "isTerminal": false,
              "value": undefined,
            },
          },
          "isTerminal": false,
          "value": undefined,
        },
      },
      "isTerminal": false,
      "value": undefined,
    },
  },
  "mapping": Map {
    "prefix" => "value",
  },
}
`;
